src/org/rust/lang/intentions/ExpandModule.kt

```diff
@@ -0,0 +1,36 @@
+package org.rust.lang.intentions
+
+import com.intellij.codeInsight.intention.IntentionAction
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.util.io.FileUtil
+import com.intellij.psi.PsiFile
+import com.intellij.psi.impl.file.PsiFileImplUtil
+import com.intellij.refactoring.move.moveFilesOrDirectories.MoveFilesOrDirectoriesUtil
+import org.rust.lang.core.psi.impl.RustFileImpl
+
+class ExpandModule : IntentionAction {
+    override fun getFamilyName(): String {
+        return "Expand module structure"
+    }
+
+    override fun startInWriteAction(): Boolean {
+        return true
+    }
+
+    override fun getText(): String {
+        return "Expand Module"
+    }
+
+    override fun isAvailable(project: Project, editor: Editor?, file: PsiFile?): Boolean {
+        return file is RustFileImpl && file.name != "mod.rs"
+    }
+
+    override fun invoke(project: Project, editor: Editor?, file: PsiFile?) {
+        val fileName = file?.name ?: return
+        val dirName = FileUtil.getNameWithoutExtension(fileName)
+        val directory = file?.parent?.createSubdirectory(dirName)
+        MoveFilesOrDirectoriesUtil.doMoveFile(file, directory)
+        PsiFileImplUtil.setName(file as PsiFile, "mod.rs")
```

Comment:
I see, but, unfortunately, Kotlin's flow sensitive type inference does not. 

You should use `file!!` to cast away `?`. I didn't know that `as` allows to cast away `?` btw ;)

Still, I would prefer an explicit check at the beginning of the function: invariant checked by the compiler tend to be more invariant.